contents/dataframes_performance.md

```diff
@@ -4,8 +4,240 @@ So far, we didn't think about making our `DataFrames.jl` code **fast**.
+
+As before, let's try out our two ways to access a column in the responses `DataFrame`:
+
+```jl
+s = """
+    # allocation # hide
+    df = responses()
+    @allocated col = df[:, :id]
+    """
+sco(s; process=string, post=plainblock)
 ```
+
+
+```jl
+s = """
+    # non allocation # hide
+    df = responses()
+    @allocated col = df[!, :id]
+    """
+sco(s; process=string, post=plainblock)
+```
+
+When we access a column without copying it we are making zero allocations and our code should be faster.
+So, if you don't need a copy always access your `DataFrame`s columns with `df.col` or `df[!, :col]` instead of `df[:, :col]`.
+
+### CSV.read versus CSV.File
+
+If you take a look at the help output for `CSV.read`, you will see that it is a convenience function identical to the function called `CSV.File` with the same keywords arguments.
+Both `CSV.read` and `CSV.File` will read the contents of a CSV file, but they differ in the default behavior.
+**`CSV.read`, by default, will not make copies** of the incoming data.
+Instead, `CSV.read` will pass all the data to the second argument (known as the "sink").
+
+So, something like this:
+
+```julia
+df = CSV.read("file.csv", DataFrame)
+```
+
+will pass all the incoming data from `file.csv` to the `DataFrame` sink, thus returning a `DataFrame` type that we store in the `df` variable.
+
+For the case of **`CSV.File`, the default behavior is the opposite: it will make copies of every column contained in the CSV file**.
+Also, the syntax is slightly different.
+We need to wrap anything that `CSV.File` returns in a `DataFrame` constructor function:
+
+```julia
+df = DataFrame(CSV.File("file.csv"))
+```
+
+Or, in a more preferred idiomatic syntax with the pipe `|>` operator:
+
+
+```julia
+df = CSV.File("file.csv") |> DataFrame
+```
+
+Like we said, `CSV.File` will make copies of each column in the underlying CSV file.
+Ultimately, if you want the most performance, you would definitely use `CSV.read` instead of `CSV.File`.
+That's why we only covered `CSV.read` in @sec:csv.
+
+### CSV.jl Multiple Files
+
+Now let's turn our attention to the `CSV.jl`.
+Specially when we have multiple CSV files to read into a single `DataFrame`.
+Since version 0.9 of `CSV.jl` we can provide a vector of strings representing filenames.
+Before, we needed to perform some sort of multiple file reading and then concatenate vertically the results into a single `DataFrame`.
+To exemplify, the code below reads from multiple CSV files and them concatenates vertically using `vcat` them into a single `DataFrame` with the `reduce` function:
+
+```julia
+files = filter(endswith(".csv"), readdir())
+df = reduce(vcat, CSV.read(file, DataFrame) for file in files)
+```
+
+One additional quirk is that `reduce` will not paralelize because it needs to keep the order of `vcat` which follows the same ordering of the `files` vector.
+
+With the new functionality in `CSV.jl` we simply pass the `files` vector into the `CSV.read` function:
```

Comment:
```suggestion
With this functionality in `CSV.jl` we simply pass the `files` vector into the `CSV.read` function:
```
soon it will not be new, just the norm.